### ISSUE TITLE
Generalized Folding Rules

### ISSUE THREAD

**Issue by PieterBranderhorst** [microsoft/vscode#138093]
I'm working with some languages where I'd like to have code folding available. Indent based folding with markers is little help because, to put it kindly, there's a lot of technical debt in that code. And some of the syntax relies on column 1. I'd rather have a parameter driven solution than write language specific providers.

I've spent some time thinking, looking at what other editors do, looking at past questions here about folding. I have a suggestion for an approach. I've written code to implement it in vscode to at least demonstrate viability. If the code isn't isn't wanted I'll work on releasing it as an extension. But like the recent bracketing work it will run best if built-in I think.

I ended up thinking that generalized folding doesn't fit perfectly with language syntax trees. Folding has considerations for line breaks, comments, blank lines. And doesn't necessarily require a language syntax - it may be useful to loosely define folding for some report text, log files, etc.

I think generic folding should support "block folding" for blocks which have a defined start and end, "heading folding" for things like chapters and sections which only have defined starts, and "group folding" where one wants consecutive lines with a particular pattern to fold as a group.

The code I've written is faster than the existing indent based folding code, way faster than the default typescript folding provider, and doesn't use much memory. It is respectful of user response time, slicing up work which could otherwise hog the cpu with very large files. (See the "TimeSlicedPromise" class in the code for that part. I didn't find something like it in vscode so I added it. It might be useful elsewhere.)

Please visit [this page](https://pieterbranderhorst.github.io/vscode-work) for more details. I hope this will be of interest.

**Comment by aeschli** [comment]
Hi @PieterBranderhorst. Thanks a lot for sharing your work. I had a glance over the code. Well crafted and nicely done!

Also FYI @alexdima and @hediet .

Is it something for VS Code? Please understand that I'm quite defensive adding new features given the maintenance work this means. We have an extension story for folding, and there's currently not much pressure to improve the built-in indentation based folding story.

The declaration based regexp approach that you chose is ideal to get started quickly. It's also suited for user-configurable folding, e.g. for custom or rarely used languages. That said, regexes can also be expensive and malformed expressions could lock the editor which is our biggest concern. Also, with real-world languages it's incredibly challenging to parse syntax correctly and get the structure of a program with regexes. The complexity of the Text Mate grammars illustrates that. That's why IMO a parser based approach is IMO a better investment.

Having your contribution as a folding provider contributed by an extension would be my preferred way. That way you can prove the concept, tune the parameters add value to users and are not blocked by us.

One limitation of the current folding providers is that if there are multiple folding providers for a given language, we try to merge the results. That's not useful in your case. We can look into:
- for the provider to discover that this is the case so it can back off (e.g. your configuration-based provider would stop if there's also language aware provider around)
- for the user to select which provider to use, like it's done for formatting
- a way to classify providers (configuration based, language extension based) that we can add to the current 'foldingStrategy' setting 

What do you think? Would that be a direction that you're interested in heading to?

**Comment by PieterBranderhorst** [comment]
Hi Martin,

Thanks for the feedback.

This code is something I hope to be for vscode. I understand you being defensive about adding stuff. I would feel the same way. I'd actually slightly dread a submission like what I sent, thinking oh no, another person who thinks they're the world's greatest programmer, just what I need. All I can say is that I hope you'll have a closer look at what I sent. Closely read some of it. There's a reason for almost every decision there, e.g. wrapping Promise instead of extending it. All debatable reasons of course. But maybe some new thoughts you'll find of interest.

I know there's not much pressure on folding. But its usage might increase if it worked better. If you end up enthused about this direction I have other thoughts which might build on it. E.g. would it be nice if foldable regions temporarily highlighted and if folded temporarily unfolded while the mouse was over the fold chevron? Add to that key bindings like we can use while navigating expandable trees in many programs (e.g. Windows Explorer) and folded code could be a nice way to navigate. Also I'm thinking that indent folding could be added to this code so that just one builtin folding module is required.

Regarding having multiple folding providers, that's something I wanted to discuss with you. For now I left a TODO in folding.ts where I changed it to invoke the new range provider. Most users would first choose a syntax/semantic language extension. Then if it also provides folding they might choose to override that. And that might be on per language basis. How best to enable that as an easy choice for users? I don't know. I saw the code which merges folding ranges from multiple providers. That was worth trying but it has some clear shortcomings. Something more controllable seems desirable.

Regarding regular expressions and their limitations I agree with what you say but I don't think the concerns are that severe. Users can already lock the editor with a bad regular expression in a search. The users who create folding rules will usually be advanced, this purpose won't often tempt them toward nasty expressions, they'll test a bit before publishing, and the documentation can steer many of them away from the problem cases. (Just never using .* or .+ except when immediately followed by $ avoids the most common traps.) Still, I know some will sneak through once in a while. In terms of handling odd cases regular expressions do of course encounter limits. But not terribly many. As an example Typescript and Javascript have a parsing nightmare in their regular expression literal syntax. The regular expression I used to handle them while testing the suggested folding code with typescript is an evil thing but does the job. (See [this file](https://github.com/PieterBranderhorst/vscode/tree/pattern_folding/extensions/typescript-basics/language-configuration.json) for the full typescript configuration. The "skip" value handles regex literals.) It can be faked out with some effort but not by any real code I've seen. There will be languages which can't be well described even for this purpose but they'll be rare. On the plus side it is easy to add new rules such as grouping "import" lines into fold groups with a parameter driven solution. And easy to add user configurable choices such as showing/hiding end of block lines. The Text Mate grammars have greater challenges than rules for folding because they need to understand a different set of things which is far more complex.

Anyway, it is all food for thought I hope. I have no desire to force anything on you. I do hope you'll take a long-shot chance that there might be more here than you ever expect in code coming out of the blue and have a close look at the code I sent. If not then let's talk about whether a change to user options to make choosing a folding approach by language is possible.

**Comment by VSCodeTriageBot** [comment]
We closed this issue because we [don't plan to address it](https://aka.ms/vscode-out-of-scope) in the foreseeable future. If you disagree and feel that this issue is crucial: we are happy to listen and to reconsider.

If you wonder what we are up to, please see our [roadmap](https://aka.ms/vscoderoadmap) and [issue reporting guidelines](https://aka.ms/vscodeissuereporting).

Thanks for your understanding, and happy coding!
